webhook/resourcesemantics/validation/validation_admit.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/webhook"
+	"knative.dev/pkg/webhook/resourcesemantics"
+)
+
+var errMissingNewObject = errors.New("the new object may not be nil")
+
+// Callback is a generic function to be called by a consumer of validation
+type Callback func(ctx context.Context, unstructured *unstructured.Unstructured, dryRun bool, opVerb admissionv1beta1.Operation) error
```

Comment:
We already have behaviour to decorate the context with the request op etc. so I'd like to be consistent and decorate the context.

Thus making the callback
```go
type Callback func(ctx context.Context, unstructured *unstructured.Unstructured) error
```

Reply:
helpers such as - `webhook.IsCreate/Update/DryRun` also would allow us to migrate to admission v1 without needing to change much code downstream

Reply:
While we do have that, that's an antipattern (API obscurity and each lookup is an O(N) operation).
Perhaps (maybe not here, but as a followup), we should reconsider and switch to explicit
```
type WHContext struct {
  context.Context
  inUpdate,dryRun bool
  ...
}
```
So it'll work both as a context and explicitly denote the params.
/cc @mattmoor

Reply:
My fear is that API implementers will forget about the dryrun scenario if it's buried in the context which could be dangerous, or ignore that the Delete verb (currently unsupported by Validatable) is super important.

Using https://medium.com/@cep21/how-to-correctly-use-context-context-in-go-1-7-8f2c0fafdf39#9e77 as a guidance:
"Just because you’ve taken that ugliness and obscured it inside Context.Value doesn’t make your API or design better. Obscurity is the opposite of good API design."

Putting important-for-control-flow values in as arguments or at least a strongly type struct helps point future implementers toward self-documenting functions whose contracts are clear.

Introducing a brand-new interface into /pkg might be a really good opportunity to reconsider this pattern.

Reply:
If you look back at commit "Merge branch 'vcontext' into unstructured" https://github.com/knative/pkg/pull/1172/commits/e1bacc8e358f95070a6f4696e98c9d792796e15f

that was my first attempt at what @vagababov is suggesting

Reply:
>Perhaps (maybe not here, but as a followup), we should reconsider and switch to explicit

Long term I agree. I would want this to be part of bigger webhook v2 rewrite that better layers these concerns.

Reply:
Should we file issues for that? So it's not forgotten until it floats up next time?

Reply:
Everything moved to the context.